### PR TITLE
Change S.R.M.TypeName to be consumed from System.Reflection.Metadata package

### LINKED ISSUE #101628: Outerloop test builds broken
Noticed this in some of the jit optional pipelines but it's in outerloop too.

https://dev.azure.com/dnceng-public/public/_build/results?buildId=657460&view=results

```
/__w/1/s/src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs(71,85): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
##[error]src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs(71,85): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj]
/__w/1/s/src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(85,31): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
##[error]src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(85,31): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj]
/__w/1/s/src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(170,50): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
##[error]src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(170,50): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj]
/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs(182,23): error CS0436: The type 'AssemblyNameInfo' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs' conflicts with the imported type 'AssemblyNameInfo' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
...etc..
```
Looks like it all may be ILVerify related.

### PR DESCRIPTION
**Pull request by jkotas**
Contributes to #101541

Fixes #101628

### CODE REVIEW

**Comment by jkotas** [comment]
This should be enough to fix the build breaks in outer loop and source build. Further cleanup is tracked by https://github.com/dotnet/runtime/issues/101541.

File: src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs [hunk 1]
```diff
@@ -8,6 +8,28 @@
...
 #nullable enable
 
+#if SYSTEM_PRIVATE_CORELIB
+namespace System.Reflection.Metadata
+{
+    internal struct TypeNameParseOptions
+    {
+        public TypeNameParseOptions() { }
+#pragma warning disable CA1822 // Mark members as static
+        // CoreLib does not enforce any limits
```
Comment (jeffhandley) [inline src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs:18]: Thanks for the comment on this; makes sense that we need the helper to pull this off more cleanly.

**Review by jeffhandley — APPROVED** [review]
Thank you, @jkotas. I appreciate you digging into this before I was able to get to it and before code flow was adversely disrupted.

**Comment by adamsitnik** [comment]
@jkotas @jeffhandley big thanks on fixing the issue when I was offline!

**Pull request merged**